modules/fedimint-wallet-client/src/lib.rs

```diff
@@ -215,6 +223,133 @@ impl ClientModuleInit for WalletClientInit {
+        /// deposits that might have been made after multiple re-derivation,
+        /// but we want also to avoid accidental address re-use.
+        const NUM_IDX_SKIP_AFTER_LAST_USED: u64 = 3;
+
+        let data = WalletClientModuleData {
+            cfg: args.cfg().clone(),
+            module_root_secret: args.module_root_secret().clone(),
+        };
+
+        let rpc_config = self
+            .0
+            .clone()
+            .unwrap_or(WalletClientModule::get_rpc_config(args.cfg()));
+
+        let btc_rpc = create_bitcoind(&rpc_config, TaskGroup::new().make_handle())?;
+
+        let mut last_used_idx = start_tweak_idx;
+        let mut cur_tweak_idx = start_tweak_idx.unwrap_or_default();
+        let mut tweak_idxes_with_pegins = BTreeSet::new();
+
+        debug!(target: LOG_CLIENT_MODULE_WALLET, max_gap=MAX_GAP, ?start_tweak_idx, "Checking history");
+
+        let new_start_tweak_idx = loop {
+            let gap_since_last_used = cur_tweak_idx - last_used_idx.unwrap_or_default();
+
+            args.update_recovery_progress(RecoveryProgress {
+                complete: u32::try_from(cur_tweak_idx.0).unwrap_or(u32::MAX),
+                total: u32::try_from(cur_tweak_idx.0.saturating_add(MAX_GAP)).unwrap_or(u32::MAX),
+            });
+
+            if MAX_GAP < gap_since_last_used {
+                break last_used_idx
+                    .unwrap_or_default()
+                    .advance(NUM_IDX_SKIP_AFTER_LAST_USED);
+            }
+
+            let (script, address, _tweak_key, _operation_id) =
+                data.derive_peg_in_script(cur_tweak_idx);
+            btc_rpc.watch_script_history(&script).await?;
+            let history = btc_rpc.get_script_history(&script).await?;
+
+            debug!(target: LOG_CLIENT_MODULE_WALLET, %cur_tweak_idx, %address, history_len=history.len(), "Checked address");
+            if !history.is_empty() {
+                last_used_idx = Some(cur_tweak_idx);
+                tweak_idxes_with_pegins.insert(cur_tweak_idx);
+            }
+            cur_tweak_idx = cur_tweak_idx.next();
+        };
+
+        let now = fedimint_core::time::now();
+        let mut dbtx = db.begin_transaction().await;
+
+        let mut tweak_idx = TweakIdx(0);
+
+        while tweak_idx < new_start_tweak_idx {
+            let (_script, _address, _tweak_key, operation_id) =
+                data.derive_peg_in_script(tweak_idx);
+            dbtx.insert_new_entry(
+                &PegInTweakIndexKey(tweak_idx),
+                &PegInTweakIndexData {
+                    creation_time: now,
+                    next_check_time: if tweak_idxes_with_pegins.contains(&tweak_idx) {
+                        // The addresses that were already used before, or didn't seem to contain
+                        // anything don't need automatic peg-in attempt, and can be re-attempted
+                        // manually if needed.
+                        Some(now)
```

Comment:
I'm thinking out loud and not sure if we need to change, but I believe this degrades privacy.

If a single client has only used Tor to claim peg-ins periodically in a mint with many Tor clients, the claims can't be correlated to a single client. If the client initiates recovery, the previously claimed peg-ins will become correlated to a single client.

Actually, now that I'm rereading the associated comment, perhaps the intention was to flip the logic? If there's already history for an address we don't schedule a next check time, but if there isn't activity we mark the tweak index due for a check now.

Reply:
During recovery client goes through a gap of 10 sequentially derived addressess and ask for their history from blockchain nodes (esplora, etc.). That already potentially correlates them in the external source that is possibly not trusted.

In this code we trigger actual claiming attempt for ones that were detected to have history. This logic will fail (gracefully) if they were already claimed, or succeed if not. That claiming itself does not change anything for the blockchain node source, though it does correlates them in the Federation.

However I don't see a good way to avoid it entirely. We need to ask Federation to check if peg-in was claimed or not. Possibly we could add some randomized delays for stuff here and in the scanning itself, but I'm not sure if it helps much, and it does have UX downsides.


We could also make the gap smaller, and the padding higher. So e.g. we only scan looking for 5 unused addresses, and use the last a new next peg-in address. The problem here is that it could lead theoretically to situation where n-th recovery is missing things because the gap between peg-in addresses is actually higher then the gap we're scanning. Though it's a very small chance.